app/src/main/java/org/projectbuendia/client/ui/OdkActivityLauncher.java

```diff
@@ -554,166 +546,19 @@ private static String readFromPath(String path) {
-        common.put(Contracts.Observations.UUID, (String) null);
-        common.put(Contracts.Observations.PATIENT_UUID, patientUuid);
-
-        final DateTime encounterTime = getEncounterAnswerDateTime(savedRoot);
-        if(encounterTime == null) return;
-        common.put(Contracts.Observations.ENCOUNTER_MILLIS, encounterTime.getMillis());
-        common.put(Contracts.Observations.ENCOUNTER_UUID, UUID.randomUUID().toString());
-
-        Set<Integer> xformConceptIds = new HashSet<>();
-        List<ContentValues> toInsert = getAnsweredObservations(common, savedRoot, xformConceptIds);
-        Map<String, String> xformIdToUuid = mapFormConceptIdToUuid(xformConceptIds, resolver);
-
-        // Remap concept ids to uuids, skipping anything we can't remap.
-        for (Iterator<ContentValues> i = toInsert.iterator(); i.hasNext(); ) {
-            ContentValues values = i.next();
-            if (!mapIdToUuid(xformIdToUuid, values, Contracts.Observations.CONCEPT_UUID)) {
-                i.remove();
-            }
-            mapIdToUuid(xformIdToUuid, values, Contracts.Observations.VALUE);
-        }
-
-        resolver.bulkInsert(Contracts.Observations.CONTENT_URI,
-            toInsert.toArray(new ContentValues[toInsert.size()]));
-    }
-
-    /** Get a map from XForm ids to UUIDs from our local concept database. */
-    private static Map<String, String> mapFormConceptIdToUuid(Set<Integer> xformConceptIds,
-                                                              ContentResolver resolver) {
-        String inClause = Joiner.on(",").join(xformConceptIds);
-
-        HashMap<String, String> xformIdToUuid = new HashMap<>();
-        Cursor cursor = resolver.query(Contracts.Concepts.CONTENT_URI,
-            new String[] {Contracts.Concepts.UUID, Contracts.Concepts.XFORM_ID},
-            Contracts.Concepts.XFORM_ID + " IN (" + inClause + ")",
-            null, null);
-
-        try {
-            while (cursor.moveToNext()) {
-                xformIdToUuid.put(Utils.getString(cursor, Contracts.Concepts.XFORM_ID),
-                    Utils.getString(cursor, Contracts.Concepts.UUID));
-            }
-        } finally {
-            cursor.close();
+    private static void updateObservationCache(JSONObject response, ContentResolver resolver) {
+
+        // TODO: don't parse this here or do a roundtrip text --> JSON --> text --> GSON conversion
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        Serializers.registerTo(gsonBuilder);
+        Gson gson = gsonBuilder.create();
+        JsonEncounter jsonEncounter = gson.fromJson(response.toString(), JsonEncounter.class);
+        Encounter encounter = Encounter.fromJson(jsonEncounter.patient_uuid, jsonEncounter);
+        ContentValues[] values = encounter.toContentValuesArray();
+        if (values.length > 0) {
+            resolver.bulkInsert(Contracts.Observations.CONTENT_URI, values);
+            
```

Comment:
Awesome refactoring, @capnfabs. It helps fixing #125 in a such cleaner and clever way. Congrats.